### PR TITLE
Make zone load balancing logic depend on local_cluster host distribution.

### PR DESCRIPTION
**Pull request by RomanDzhabarov**
Documentation around runtime control/stats/general overview are going to be on a separate commit.

I was running simulation tests from load_balancer_simulation_test.cc, results look very good.

### CODE REVIEW

File: source/common/upstream/load_balancer_impl.cc [hunk 1]
```diff
@@ -46,48 +59,92 @@ bool LoadBalancerBase::isGlobalPanic() {
...
+const std::vector<HostPtr>& LoadBalancerBase::tryChooseLocalZoneHosts() {
   // At this point it's guaranteed to be at least 2 zones.
-  uint32_t number_of_zones = host_set_.healthyHostsPerZone().size();
-  ASSERT(number_of_zones >= 2U);
-  const std::vector<HostPtr>& local_zone_healthy_hosts = host_set_.healthyHostsPerZone()[0];
+  ASSERT(host_set_.healthyHostsPerZone().size() >= 2U);
 
-  // If number of hosts in a local zone big enough then route all requests to the same zone.
-  if (local_zone_healthy_hosts.size() * number_of_zones >= host_set_.healthyHosts().size()) {
+  std::vector<uint64_t> local_percentage =
```
Comment (mattklein123) [inline source/common/upstream/load_balancer_impl.cc]: Use a stack allocated array with the right size, rework the functions above to take the array and populate it if you want to do that. You can't allocate things like vectors in this code path.

Comment (mattklein123) [inline source/common/upstream/load_balancer_impl.cc]: Or just do the pre-calculation work now. I would actually prefer you just do that.

File: source/common/upstream/load_balancer_impl.cc [hunk 2]
```diff
@@ -46,48 +59,92 @@ bool LoadBalancerBase::isGlobalPanic() {
...
   }
 
-  // If local zone ratio is lower than expected we should only partially route requests from the
-  // same zone.
-  double zone_host_ratio = 1.0 * local_zone_healthy_hosts.size() / host_set_.healthyHosts().size();
-  double ratio_to_route = zone_host_ratio * number_of_zones;
+  // At this point we should route cross zone as we cannot route to the local zone.
+  stats_.zone_routing_no_sampled_.inc();
+
+  std::vector<uint64_t> capacity_left;
```
Comment (mattklein123) [inline source/common/upstream/load_balancer_impl.cc]: stack allocated array

File: source/common/upstream/load_balancer_impl.cc [hunk 3]
```diff
@@ -46,48 +59,92 @@ bool LoadBalancerBase::isGlobalPanic() {
...
+  // Local zone does not have additional capacity (we already routed what we could), but
```
Comment (mattklein123) [inline source/common/upstream/load_balancer_impl.cc]: More comments. Don't understand what is happening here.

File: source/common/upstream/load_balancer_impl.cc (hunk 2 above)
Comment (RomanDzhabarov) [inline source/common/upstream/load_balancer_impl.cc]: fixed

File: source/common/upstream/load_balancer_impl.cc (hunk 3 above)
Comment (RomanDzhabarov) [inline source/common/upstream/load_balancer_impl.cc]: fixed

File: source/common/upstream/load_balancer_impl.cc (hunk 1 above)
Comment (RomanDzhabarov) [inline source/common/upstream/load_balancer_impl.cc]: i made stack alloc for now, i want make sure things works good in prod

File: source/common/upstream/load_balancer_impl.cc [hunk 4]
```diff
@@ -46,48 +59,102 @@ bool LoadBalancerBase::isGlobalPanic() {
...
+  }
+
+  // At this point we must route cross zone as we cannot route to the local zone.
+  stats_.zone_routing_cross_zone_.inc();
+
+  // Local zone does not have additional capacity (we have already routed what we could).
+  // Now we need to figure out how much traffic we can route cross zone and to which exact zone
+  // we should route. Percentage of requests routed cross zone to a specific zone should be
+  // proportional to the residual capacity upstream zone has.
+  //
```
Comment (ccaraman) [inline source/common/upstream/load_balancer_impl.cc]: Needed?

Comment (RomanDzhabarov) [inline source/common/upstream/load_balancer_impl.cc]: fixed

File: include/envoy/upstream/upstream.h [hunk 5]
```diff
@@ -192,10 +192,12 @@ class HostSet {
   COUNTER(update_success)                                                                          \
   COUNTER(update_failure)                                                                          \
   COUNTER(zone_cluster_too_small)                                                                  \
-  COUNTER(zone_over_percentage)                                                                    \
+  COUNTER(zone_routing_all_directly)                                                               \
```
Comment (mattklein123) [inline include/envoy/upstream/upstream.h]: nit: can you prefix all of the lb ones with "lb_" and then reorder this list.

Comment (RomanDzhabarov) [inline include/envoy/upstream/upstream.h]: yep, will do for consistency with panic threshold

File: source/common/upstream/load_balancer_impl.cc [hunk 6]
```diff
@@ -46,48 +59,102 @@ bool LoadBalancerBase::isGlobalPanic() {
...
+
   ASSERT(number_of_zones >= 2U);
-  const std::vector<HostPtr>& local_zone_healthy_hosts = host_set_.healthyHostsPerZone()[0];
 
-  // If number of hosts in a local zone big enough then route all requests to the same zone.
-  if (local_zone_healthy_hosts.size() * number_of_zones >= host_set_.healthyHosts().size()) {
-    stats_.zone_over_percentage_.inc();
-    return local_zone_healthy_hosts;
+  uint64_t local_percentage[number_of_zones];
+  calculateZonePercentage(local_host_set_->healthyHostsPerZone(), local_percentage);
```
Comment (mattklein123) [inline source/common/upstream/load_balancer_impl.cc]: ASSERT here also that local_host_set_->healthyHostsPerZone().size() == host_set_.healthyHostsPerZone().size()

File: source/common/upstream/load_balancer_impl.h [hunk 7]
```diff
@@ -26,8 +26,26 @@ class LoadBalancerBase {
...
   bool earlyExitNonZoneRouting();
-  bool isGlobalPanic();
+
+  /**
+   * For the given host_set it @return if we should be in a panic mode or not.
+   * For example, if majority of hosts are unhealthy we'll be likely in a panic mode.
+   * In this case we'll route requests to hosts no matter if they are healthy or not.
+   */
+  bool isGlobalPanic(const HostSet& host_set);
+  const std::vector<HostPtr>& tryChooseLocalZoneHosts();
```
Comment (mattklein123) [inline source/common/upstream/load_balancer_impl.h]: nit: add comment or newline before this line

File: source/common/upstream/load_balancer_impl.cc [hunk 8]
```diff
@@ -46,48 +59,102 @@ bool LoadBalancerBase::isGlobalPanic() {
...
+  // proportional to the residual capacity upstream zone has.
+  //
+  // residual_capacity contains capacity left in a given zone, we keep accumulating residual
+  // capacity to make search for sampled value easier.
+  uint64_t residual_capacity[number_of_zones];
+
+  // Local zone (index 0) does not have residual capacity as we have routed all we could.
+  residual_capacity[0] = 0;
+  for (size_t i = 1; i < number_of_zones; ++i) {
+    // Only route to the zones that have additional capacity.
```
Comment (mattklein123) [inline source/common/upstream/load_balancer_impl.cc]: still having trouble quickly understanding this logic. More comments, and potentially an example.

File: source/common/upstream/load_balancer_impl.cc [hunk 9]
```diff
@@ -46,48 +59,102 @@ bool LoadBalancerBase::isGlobalPanic() {
...
+      // Zone with index "i" does not have residual capacity, but we keep accumulating previous
+      // values to make search easier on the next step.
+      residual_capacity[i] = residual_capacity[i - 1];
+    }
+  }
 
-  // Not zone routed requests will be distributed between all hosts and hence
-  // we need to route only fraction of req_percent_to_route to the local zone.
-  double actual_routing_ratio = (ratio_to_route - zone_host_ratio) / (1 - zone_host_ratio);
+  // Random sampling to select specific zone for cross zone traffic based on the additional
```
Comment (mattklein123) [inline source/common/upstream/load_balancer_impl.cc]: more comments. What type of number does residual_capacity[number_of_zones - 1] have in it. What range?

File: source/common/upstream/load_balancer_impl.cc (hunk 6 above)
Comment (RomanDzhabarov) [inline source/common/upstream/load_balancer_impl.cc]: added

File: source/common/upstream/load_balancer_impl.h (hunk 7 above)
Comment (RomanDzhabarov) [inline source/common/upstream/load_balancer_impl.h]: put comments.

File: source/common/upstream/load_balancer_impl.cc (hunk 9 above)
Comment (RomanDzhabarov) [inline source/common/upstream/load_balancer_impl.cc]: more comments added

File: source/common/upstream/load_balancer_impl.cc (hunk 8 above)
Comment (RomanDzhabarov) [inline source/common/upstream/load_balancer_impl.cc]: more comments added

File: source/common/upstream/load_balancer_impl.cc [hunk 10]
```diff
@@ -25,69 +25,147 @@ bool LoadBalancerBase::earlyExitNonZoneRouting() {
...
+  // capacity to make search for sampled value easier.
+  // For example, if we have the following upstream and local percentage:
+  // local_percentage: 40000 40000 20000
+  // upstream_percentage: 25000 50000 25000
+  // Residual capacity would look like: 0 10000 5000. Now we need to sample proportionally to
+  // bucket sizes (residual capacity). For simplicity of finding where specific
+  // sampled value is, we accumulate values in residual capacity. This is what it will look like:
+  // residual_capacity: 0 10000 15000
+  // Now to find zone to route (bucket) we could simply iterate over residual_capacity searching
+  // where
```
Comment (mattklein123) [inline source/common/upstream/load_balancer_impl.cc]: nit: can you fix line wrapping in this comment block. Keep to 100 col flow unless you need a new paragraph.

Comment (RomanDzhabarov) [inline source/common/upstream/load_balancer_impl.cc]: :( fix format broke this... fixing

**Pull request merged**